### ISSUE TITLE
Support Postgres Interval in GraphQL

### ISSUE THREAD

**Issue by Joey92** [directus/directus#12550]
### Preflight Checklist

- [X] I have completed all [Troubleshooting Steps](https://docs.directus.io/getting-started/support/#troubleshooting-steps).
- [X] I'm on [the latest version of Directus](https://github.com/directus/directus/releases).
- [X] There's [no other issue](https://github.com/directus/directus/issues?q=is%3Aissue) that already describes my problem.

### Describe the Bug

When I want to access fields via GraphQL that are of type interval in Postgres the values are null.
In the GraphQL errors I get a message like this: `String cannot represent value: { hours: 22, minutes: 36, seconds: 55, milliseconds: 220 }` 
When querying it as a REST call it works fine.

I guess the pg client for node gives you the data as JSON, but GraphQL takes the type of the field configured in Directus, which seems to be treating it as a string. (My educated guess here)

### To Reproduce

Create a field in Postgres of type Interval, not via Directus admin UI. Add some data and query the collection via GraphQL.

### Errors Shown

Browser console: String cannot represent value: { hours: 22, minutes: 36, seconds: 55, milliseconds: 220 }

### What version of Directus are you using?

9.8.0

### What version of Node.js are you using?

16.9.1

### What database are you using?

Postgres 13.4

### What browser are you using?

Firefox

### What operating system are you using?

Linux

### How are you deploying Directus?

As project dependency in package.json, runtime in node16-alpine docker container

**Comment by azrikahar** [comment]
> I guess the pg client for node gives you the data as an object, but GraphQL takes the type of the field configured in Directus, which seems to be treating it as a string. (My educated guess here)

You're right on the mark here! Currently `interval` is treated/"mapped" as `string` by default:

https://github.com/directus/directus/blob/703fb8426f59619a4f21e80f670ed4347d01666c/api/src/utils/get-local-type.ts#L84

and then for GraphQL, a util function maps the field type to the appropriate GraphQL type, but in this case, it defaults to `GraphQLString`:

https://github.com/directus/directus/blob/703fb8426f59619a4f21e80f670ed4347d01666c/api/src/utils/get-graphql-type.ts#L14-L37

---

Changing the 1st referenced code to `json` does resolve your issue, but I'm wondering how can we make the experience better on the app side, since treating it as json ended up showing the code interface:

![image](https://user-images.githubusercontent.com/42867097/161488610-360b93fb-812c-4ebb-b552-97d9b2cdd437.png)

Although on surface it seems like we can edit it just fine like a JSON, but it is in fact not and should not be treated as such (or else we'll have weird results), as stated by this SO answer: https://stackoverflow.com/a/58876854. Here's the "weirdness" in action, where hour gets saved as minutes instead:

https://user-images.githubusercontent.com/42867097/161489300-699fe9f4-a564-4071-904e-51571142c1fa.mp4

Saving it as a string like "2h 30m" despite it is a code interface does work! However there's also currently the issue #11098 that is disrupting this workaround at the moment.

---

@Joey92 this isn't ideal and more so a temporary workaround, but can you try going to `directus_fields` and update the `special` column for your interval field row to `cast-json`? It should make it work for both REST and GraphQL, just that the value editing experience on the App side is not perfect.

**Comment by Joey92** [comment]
Wow thank you for your detailed response!

Treating the field as a string is actually not that bad. It's an object because of the node pg client. In the database itself you can also insert into an interval field as a string `30:00`. There are a bunch of ways the database could be fed by casting a string to interval on query level: https://www.postgresql.org/docs/current/datatype-datetime.html#DATATYPE-INTERVAL-INPUT-EXAMPLES

I'll try out your suggestion setting the field to `cast-json` later today. It's a read-only field in my app, so no need to worry about the editing experience. :)

**Comment by rijkvanzanten** [comment]
This is _technically_ more of a feature request to properly support Interval types in the platform, but lets first treat it as a bug to make sure it's fallback support is returned consistently across both rest and graphql 🙂

**Comment by Joey92** [comment]
@rijkvanzanten Yep I was really unsure on where to put it. I choose Bug since it's technically working with REST but not with GraphQL

@azrikahar Unfortunately I get the same error as before. Directus does now show my field as type JSON in the UI though.

**Comment by Joey92** [comment]
@azrikahar thanks for checking for a workaround. For now I'm using the REST endpoint for the that specific call, instead of GraphQL. Keeping the response as an object is helpful because my date time lib can directly use it to manipulate the DateTime object.

One could argue though if the value of the field should be represented as an object instead of a scalar value. But that is not my decision to make.

**Comment by rijkvanzanten** [comment]
> This is _technically_ more of a feature request to properly support Interval types in the platform, but lets first treat it as a bug to make sure it's fallback support is returned consistently across both rest and graphql 🙂

I thiiiiiiink I'm circling back on this statement. We're not officially supporting interval values in any capacity, which means that the REST output is simply whatever the `pg` library gives us through `knex`. The only difference between REST and GraphQL is that the REST implementation is a little leaner in unexpected type signatures, and will return whatever the database spits out. GraphQL on the other hand validates both input and output, and will throw the (expected) error that the type doesn't match what Directus expects (which is correct, as Directus hasn't been built to expect intervals in the first place). Moving this to a feature request discussion!
